pkg/metadata/scheduler.go

```diff
@@ -74,20 +87,42 @@ func (c *Scheduler) AddCollector(name string, interval time.Duration) error {
 			select {
 			case <-ctx.Done():
 				return
-			case <-health.C:
-			case <-sendTicker.C:
+			case <-sc.healthHandle.C:
+				// Purposely empty
+			case <-sc.sendTimer.C:
+				sc.sendTimer.Reset(interval) // Reset the timer, so it fires again after `interval`.
+				// Note we call `p.Send` on the collector *after* resetting the Timer, so
+				// the time spent by `p.Send` is not added to the total time between runs.
 				if err := p.Send(c.srl); err != nil {
 					log.Errorf("Unable to send '%s' metadata: %v", name, err)
 				}
 			}
 		}
 	}()
-	c.tickers = append(c.tickers, sendTicker)
-	c.healthHandles = append(c.healthHandles, health)
+	c.collectors[name] = sc
 
 	return nil
 }
 
+// SendNow runs a collector immediately and resets its ticker.
+// Does nothing if the Scheduler has been stopped, since the
+// goroutine that listens on the Timer will not be running.
+func (c *Scheduler) SendNow(name string) {
+	sc, found := c.collectors[name]
+
+	if !found {
+		log.Errorf("Unable to find '" + name + "' metadata collector in the catalog!")
+	}
+
+	if !sc.sendTimer.Stop() {
```

Comment:
From the docs, draining the channel _"... cannot be done concurrent to other receives from the Timer's channel."_ Maybe we're fine here, but it's made me think (and I may be overthinking).

wondering if we could potentially have a tiny race condition here:
1. the timer fires
2. it's consumed here: https://github.com/DataDog/datadog-agent/pull/4166/files#diff-f6baa386e7bcc8bf9d64dc79cb19be42R92, but hypothetically the go routine get's preempted before finishing the case switch, in this case before resetting the timer.
3. This `sc.sendTime.Stop()` is called, it returns `false` as the timer has expired but not yet been `Reset()` as we were preempted in (2) above.
4. We then try to drain the channel, but it was already drained in (1), so we block on `SendNow()`. 

I think this might eventually fix itself, because the timer would be reset, and maybe we'd eventually get out of that lock, but we'd be competing for that `<-sc.sendTime.C` channel read.

Given that the interval is relatively large, wondering if it might be best to avoid draining the channel? 🤔

Reply:
Yup, this could potentially block. I've fixed this by adding a `select` statement, so we continue if there is nothing to be read from the channel. 

With the new code we will never block on `SendNow`, but we would be calling `Send` twice if this scenario ever happens. I think it's a better situation, though.